dotcom-rendering/configs/webpack/server.dev.mjs

```diff
@@ -10,18 +10,114 @@
 // THIS IS ONE IS TRICKIER, THIS IS MORE OF A GUIDE FOR HOW TO PROCEED.
 // IT'S NOT FINISHED.
 
+import { resolve } from 'node:path';
+import { fileURLToPath } from 'node:url';
+import bodyParser from 'body-parser';
+// eslint-disable-next-line import/no-named-as-default -- this is the Webpack way
+import webpack from 'webpack';
+import webpackHotServerMiddleware from 'webpack-hot-server-middleware';
 import { mergeWithRules } from 'webpack-merge';
 import nodeExternals from 'webpack-node-externals';
+import { success } from '../../../scripts/log.js';
+import { getContentFromURLMiddleware } from '../../src/server/lib/get-content-from-url.js';
 import { base } from './base.mjs';
 import { server } from './server.mjs';
 
+const DIRNAME = fileURLToPath(new URL('.', import.meta.url));
+
 /** @type {import("webpack").Configuration} */
 export const devServer = {
+	devServer: {
+		port: 3030,
+		compress: false,
+		hot: false,
+		liveReload: true,
+		client: {
+			logging: 'warn',
+			overlay: true,
+		},
+		static: {
+			directory: resolve(DIRNAME, '..', '..', 'src', 'static'),
+			publicPath: '/static/frontend',
+		},
+		allowedHosts: ['r.thegulocal.com'],
+		devMiddleware: {
+			publicPath: '/assets/',
+			serverSideRender: true,
+			headers: (req, res) => {
+				// Allow any localhost request from accessing the assets
+				if (
+					req.hostname === (process.env.HOSTNAME || 'localhost') &&
+					req.headers.origin
+				)
+					res.setHeader(
+						'Access-Control-Allow-Origin',
+						req.headers.origin,
+					);
+			},
+		},
+		setupMiddlewares: (middlewares, { app, compiler }) => {
+			if (!app) {
+				throw new Error('webpack-dev-server is not defined');
+			}
+
+			// it turns out webpack dev server is just an express server
+			// with webpack-dev-middleware, so here we add some other middlewares
+			// of our own
+
+			app.use(bodyParser.json({ limit: '10mb' }));
+
+			// populates req.body with the content data from a production
+			// URL if req.params.url is present
+			app.use(getContentFromURLMiddleware);
+
+			app.get('/', (req, res) => {
+				res.sendFile(
+					resolve(
+						DIRNAME,
+						'..',
+						'..',
+						'src',
+						'server',
+						'dev-index.html',
+					),
+				);
+			});
+
+			// webpack-hot-server-middleware needs to run after webpack-dev-middleware
+			if (compiler instanceof webpack.MultiCompiler) {
+				middlewares.push({
+					name: 'server',
+					middleware: webpackHotServerMiddleware(compiler, {
+						chunkName: 'server',
+					}),
+				});
+			}
+
+			return middlewares;
+		},
+		onListening: ({ options: { port } }) => {
+			if (typeof port !== 'number') return;
+			success(
+				[
+					'DEV server running on ',
+					'\x1b[36m', // cyan
+					'\x1b[4m', // underline
+					`http://localhost:${port}`,
+					'\x1b[0m', // reset
+				].join(''),
+			);
+		},
```

Comment:
This is new and replaces the inline log, which has the added benefit of actually only showing a clickable link once the server is actually listening.